system/jlib/jstats.cpp

```diff
@@ -1974,43 +1973,52 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
     virtual IStatisticCollection * ensureSubScope(const StatsScopeId & search, bool hasChildren)
     {
         bool wasCreated;
-        return ensureSubScope(search, hasChildren, &wasCreated);
+        return ensureSubScope(search, hasChildren, wasCreated);
     }
-    virtual IStatisticCollection * ensureSubScope(const StatsScopeId & search, bool hasChildren, bool * wasCreated) override
+    virtual IStatisticCollection * ensureSubScope(const StatsScopeId & search, bool hasChildren, bool & wasCreated) override
     {
         //Once the CStatisticCollection is created it should not be replaced - so that returned pointers remain valid.
-        *wasCreated = false;
+        wasCreated = false;
         IStatisticCollection * match = children.find(&search);
         if (match)
             return match;
 
         IStatisticCollection * ret = new CStatisticCollection(this, search);
         children.add(*ret);
-        *wasCreated=true;
+        wasCreated = true;
         return ret;
     }
+    virtual IStatisticCollection * querySubScope(const StatsScopeId & search) override
+    {
+        return children.find(&search);
+    }
 
-    virtual IStatisticCollection * ensureSubScopePath(std::initializer_list<StatsScopeId> path, bool * wasCreated) override
+    virtual IStatisticCollection * ensureSubScopePath(std::initializer_list<const StatsScopeId> path, bool & wasCreated) override
     {
         IStatisticCollection * curScope = this;
-        // interested only in the 'wasCreated' for the deepest scope (i.e. the only updated last)
-        for (auto it = path.begin(); it != path.end(); ++it)
-            curScope = curScope->ensureSubScope(*it, true, wasCreated); // n.b. this will always return a valid pointer
+        for (const auto & scopeItem: path)
+            curScope = curScope->ensureSubScope(scopeItem, true, wasCreated); // n.b. this will always return a valid pointer
         return curScope;
     }
 
-    virtual IStatisticCollection * querySubScopePath(std::initializer_list<StatsScopeId> path) override
+    virtual IStatisticCollection * querySubScopePath(std::initializer_list<const StatsScopeId> path) override
     {
         IStatisticCollection * curScope = this;
-        for (auto it = path.begin(); it != path.end(); ++it)
+        for (const auto & scopeItem: path)
         {
-            curScope = children.find(&(*it));
+            curScope = curScope->querySubScope(scopeItem);
             if (!curScope)
-                return nullptr;
+                break;
```

Comment:
this doesn't seem right.
This break; (and falling through to return curScope) means as it's traversing the StatsScopeId list, if one isn't found it will return the IStatisticCollection level of wherever it happened to get to when it stopped.

It seemed right as it was, to return null if it can't satisfy the full query.

Reply:
It only breaks if curScope is null, so if it break here the return will be null.  I could return a null pointer here but I don't think it makes much difference, does  it?

(I did it this way because when I was debugging, it was easier to add debug just before a single return line)

Reply:
I misread